assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml

```diff
@@ -81,6 +173,78 @@ spec:
         - name: ssl-certs-host
           mountPath: /etc/ssl/certs
           readOnly: true
+      {{- if eq (int .Values.apiserver.replicas) 1 }}
+        - name: data
+          mountPath: /run/kube-apiserver
+      - name: haproxy
+        image: haproxy:2.1.3-alpine
+        volumeMounts:
+        - name: data
+          mountPath: /run/kube-apiserver
+        command:
+        - /bin/sh
+        - -c
+        - |
+          set -xe
+          export ADDRESS=$(cat /run/kube-apiserver/address)
+          # Make sure initContainer generated kube-apiserver address.
+          if [ -z $ADDRESS ]; then
+            echo "ADDRESS not found"
+            exit 1
+          fi
+          echo "Connecting to $ADDRESS:{{ template "port" . }}"
+          # We use TCP readiness probe and HAProxy does not reject connections if no backend is available,
+          # so we wait until kube-apiserver is available here, so readiness of haproxy container represents
+          # readiness of kube-apiserver, as kube-apiserver cannot have readiness probe set, as it listens
+          # on random IP address.
+          until nc -zv $ADDRESS {{ template "port" . }}; do sleep 1; done
+          echo "Connected"
+          # From https://github.com/docker-library/haproxy/blob/master/Dockerfile-debian.template#L70
+          exec haproxy -f /run/kube-apiserver/haproxy.cfg
+        readinessProbe:
```

Comment:
I see only the non HA case has a readinessProbe? According to https://github.com/kinvolk/lokomotive/pull/32#discussion_r390459035 don't we want to add one in the HA case too? If that is out of scope and tracked in some other place, fine by me. But if we can add a commit with this very same readinessProbe in the PR, seems on topic for me too

Reply:
I think we can add it for HA case too to make upgrade process smoother.

Reply:
I added it, but didn't test yet.

Reply:
Thanks!

Reply:
Tested now. Seems to work fine, though it doesn't play very well with pod checkpointer, as it kicks in right after the pod is descheduled when upgrading the DaemonSet.

Reply:
Ohh, the readinessProbe doesn't bode well? So maybe just removing it is better?

Reply:
It's better with `readinessProbe`, as it then upgrade process waits until new `kube-apiserver` is running correctly.

Reply:
I wouldn't block the PR on this, create an issue and we can fix later.

Reply:
Agreed. At this point, it seems better to merge and improve later for most of the remaining things (as long as basic tests we care about work fine, of course :))